From d5d2499eb4f9acd6c5127df118e8e35215323bed Mon Sep 17 00:00:00 2001
From: Stanislav Fomichev <sdf@google.com>
Date: Tue, 25 Jan 2022 16:13:40 -0800
Subject: [PATCH 351/921] [Backport] bpf: fix register_btf_kfunc_id_set for
 !CONFIG_DEBUG_INFO_BTF

mainline inclusion
from mainline-v5.18-rc1
commit c446fdacb10dcb3b9a9ed3b91d91e72d71d94b03
category: bugfix
bugzilla: N/A

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c446fdacb10dcb3b9a9ed3b91d91e72d71d94b03

------------------------------------------------------

Commit dee872e124e8 ("bpf: Populate kfunc BTF ID sets in struct btf")
breaks loading of some modules when CONFIG_DEBUG_INFO_BTF is not set.
register_btf_kfunc_id_set returns -ENOENT to the callers when
there is no module btf. Let's return 0 (success) instead to let
those modules work in !CONFIG_DEBUG_INFO_BTF cases.

Acked-by: Kumar Kartikeya Dwivedi <memxor@gmail.com>
Fixes: dee872e124e8 ("bpf: Populate kfunc BTF ID sets in struct btf")
Signed-off-by: Stanislav Fomichev <sdf@google.com>
Link: https://lore.kernel.org/r/20220126001340.1573649-1-sdf@google.com
Signed-off-by: Alexei Starovoitov <ast@kernel.org>

Conflicts:
	kernel/bpf/btf.c

Signed-off-by: Yang Jihong <yangjihong1@huawei.com>
---
 kernel/bpf/btf.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/kernel/bpf/btf.c b/kernel/bpf/btf.c
index 1062493348bc..52951ff518d0 100644
--- a/kernel/bpf/btf.c
+++ b/kernel/bpf/btf.c
@@ -5822,8 +5822,15 @@ int register_btf_kfunc_id_set(enum bpf_prog_type prog_type,
 		return -ENOTSUPP;
 
 	btf = bpf_get_btf_vmlinux();
-	if (IS_ERR_OR_NULL(btf))
-		return btf ? PTR_ERR(btf) : -ENOENT;
+	if (!btf) {
+		if (!kset->owner && IS_ENABLED(CONFIG_DEBUG_INFO_BTF)) {
+			pr_err("missing vmlinux BTF, cannot register kfuncs\n");
+			return -ENOENT;
+		}
+		return 0;
+	}
+	if (IS_ERR(btf))
+		return PTR_ERR(btf);
 
 	hook = bpf_prog_type_to_kfunc_hook(prog_type);
 	ret = btf_populate_kfunc_set(btf, hook, kset);
-- 
2.17.1

